In [26]:
import pandas as pd
import numpy as np

from geopy.geocoders import Nominatim
#from geopy.geocoders import ArcGIS

import requests
import warnings
warnings.filterwarnings('ignore')



# Lets open CSV to take Engineer the features of MRTS and CBD Distence to the property for accurate Predictions

In [27]:
df_forfeatureEng = pd.read_csv("csv_data\\csv_df_Add_2_MRTS_CBD.csv")

In [28]:
df_forfeatureEng

month        town  flat_type block       street_name storey_range  \
0       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4       1990-01  ANG MO KIO     3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   
...         ...         ...        ...   ...               ...          ...   
931258  2014-12      YISHUN     5 ROOM   816      YISHUN ST 81     10 TO 12   
931259  2014-12      YISHUN  EXECUTIVE   325       YISHUN CTRL     10 TO 12   
931260  2014-12      YISHUN  EXECUTIVE   618    YISHUN RING RD     07 TO 09   
931261  2014-12      YISHUN  EXECUTIVE   277      YISHUN ST 22     07 TO 09   
931262  2014-12      YISHUN  EXECUTIVE   277      YISHUN ST 22     04 TO 06   

        floor_area_sqm      flat_model  lease_commence_date  resale_price  \
0                 31.0        IMPROVED                 1977        9000.0   
1                 31.0        IMPROVED                 1977        6000.0   
2                 31.0        IMPROVED                 1977        8000.0   
3                 31.0        IMPROVED                 1977        6000.0   
4                 73.0  NEW GENERATION                 1976       47200.0   
...                ...             ...                  ...           ...   
931258           122.0        Improved                 1988      580000.0   
931259           146.0      Maisonette                 1988      540000.0   
931260           164.0       Apartment                 1992      738000.0   
931261           152.0      Maisonette                 1985      592000.0   
931262           146.0      Maisonette                 1985      545000.0   

       remaining_lease               address  
0                  NaN  309 ANG MO KIO AVE 1  
1                  NaN  309 ANG MO KIO AVE 1  
2                  NaN  309 ANG MO KIO AVE 1  
3                  NaN  309 ANG MO KIO AVE 1  
4                  NaN  216 ANG MO KIO AVE 1  
...                ...                   ...  
931258             NaN      816 YISHUN ST 81  
931259             NaN       325 YISHUN CTRL  
931260             NaN    618 YISHUN RING RD  
931261             NaN      277 YISHUN ST 22  
931262             NaN      277 YISHUN ST 22  

[931263 rows x 12 columns]

# Featuring Engineering - Adding Extra Features of MRTS Rail Station data and Centre of Bussiness Development Landmark proximity to the property 

In [29]:

# source from https://mrtmapsingapore.com/mrt-stations-singapore/ 

# this list of Stations are restricted to scope of the project requirement
list_of_mrt = [
    'Jurong East MRT Station',
    'Bukit Batok MRT Station',
    'Bukit Gombak MRT Station',
    'Choa Chu Kang MRT Station',
    'Yew Tee MRT Station',
    'Kranji MRT Station',
    'Marsiling MRT Station',
    'Woodlands MRT Station',
    'Admiralty MRT Station',
    'Sembawang MRT Station',
    'Canberra MRT Station',
    'Yishun MRT Station',
    'Khatib MRT Station',
    'Yio Chu Kang MRT Station',
    'Ang Mo Kio MRT Station',
    'Bishan MRT Station',
    'Braddell MRT Station',
    'Toa Payoh MRT Station',
    'Novena MRT Station',
    'Newton MRT Station',
    'Orchard MRT Station',
    'Somerset MRT Station',
    'Dhoby Ghaut MRT Station',
    'City Hall MRT Station',
    'Raffles Place MRT Station',
    'Marina Bay MRT Station',
    'Marina South Pier MRT Station',
    'Pasir Ris MRT Station',
    'Tampines MRT Station',
    'Simei MRT Station',
    'Tanah Merah MRT Station',
    'Bedok MRT Station',
    'Kembangan MRT Station',
    'Eunos MRT Station',
    'Paya Lebar MRT Station',
    'Aljunied MRT Station',
    'Kallang MRT Station',
    'Lavender MRT Station',
    'Bugis MRT Station',
    'Tanjong Pagar MRT Station',
    'Outram Park MRT Station',
    'Tiong Bahru MRT Station',
    'Redhill MRT Station',
    'Queenstown MRT Station',
    'Commonwealth MRT Station',
    'Buona Vista MRT Station',
    'Dover MRT Station',
    'Clementi MRT Station',
    'Chinese Garden MRT Station',
    'Lakeside MRT Station',
    'Boon Lay MRT Station',
    'Pioneer MRT Station',
    'Joo Koon MRT Station',
    'Gul Circle MRT Station',
    'Tuas Crescent MRT Station',
    'Tuas West Road MRT Station',
    'Tuas Link MRT Station',
    'Expo MRT Station',
    'Changi Airport MRT Station',
    'HarbourFront MRT Station',
    'Chinatown MRT Station',
    'Clarke Quay MRT Station',
    'Little India MRT Station',
    'Farrer Park MRT Station',
    'Boon Keng MRT Station',
    'Potong Pasir MRT Station',
    'Woodleigh MRT Station',
    'Serangoon MRT Station',
    'Kovan MRT Station',
    'Hougang MRT Station',
    'Buangkok MRT Station',
    'Sengkang MRT Station',
    'Punggol MRT Station',
    'Bras Basah MRT Station',
    'Esplanade MRT Station',
    'Promenade MRT Station',
    'Nicoll Highway MRT Station',
    'Stadium MRT Station',
    'Mountbatten MRT Station',
    'Dakota MRT Station',
    'MacPherson MRT Station',
    'Tai Seng MRT Station',
    'Bartley MRT Station',
    'Lorong Chuan MRT Station',
    'Marymount MRT Station',
    'Caldecott MRT Station',
    'Botanic Gardens MRT Station',
    'Farrer Road MRT Station',
    'Holland Village MRT Station',
    'one-north MRT Station',
    'Kent Ridge MRT Station',
    'Haw Par Villa MRT Station',
    'Pasir Panjang MRT Station',
    'Labrador Park MRT Station',
    'Telok Blangah MRT Station',
    'Bayfront MRT Station',
    'Bukit Panjang MRT Station',
    'Cashew MRT Station',
    'Hillview MRT Station',
    'Beauty World MRT Station',
    'King Albert Park MRT Station',
    'Sixth Avenue MRT Station',
    'Tan Kah Kee MRT Station',
    'Stevens MRT Station',
    'Rochor MRT Station',
    'Downtown MRT Station',
    'Telok Ayer MRT Station',
    'Fort Canning MRT Station',
    'Bencoolen MRT Station',
    'Jalan Besar MRT Station',
    'Bendemeer MRT Station',
    'Geylang Bahru MRT Station',
    'Mattar MRT Station',
    'Ubi MRT Station',
    'Kaki Bukit MRT Station',
    'Bedok North MRT Station',
    'Bedok Reservoir MRT Station',
    'Tampines West MRT Station',
    'Tampines East MRT Station',
    'Upper Changi MRT Station',
    'Woodlands North MRT Station',
    'Woodlands South MRT Station'
]

In [30]:
#Mrts Lat and Long Value Extraction from geoPy module
# the code block is comented to prevent the recursive scraping of geodata
'''
mrts_lat = []
mrts_long = []
# Initialize Nominatim API
geolocator=ArcGIS()
#geolocator = Nominatim(user_agent="MyApp")
#geolocator = Nominatim(user_agent="geoapiExercises")
#geolocator = Nominatim(user_agent="Geopy Library")
for i in range(0, len(list_of_mrt)):
    query_address = list_of_mrt[i]
    query_addressSplit = query_address.split(" ")
    address = "+".join(query_addressSplit)
    location = geolocator.geocode(address)
    try:
        if location.latitude < 2.0 and location.latitude > 1.0:
            formattedLat = "{:.6f}".format(location.latitude)
        else:
            formattedLat =0.0
        if location.longitude > 102.0 and location.longitude < 104.0 :
            formattedLon = "{:.6f}".format(location.longitude)
        else:
            formattedLon =0.0
        
        print (str(query_address)+",Lat: "+str(formattedLat) +" Long: "+str(formattedLon))
        mrts_lat.append(float(formattedLat))
        mrts_long.append(float(formattedLon))

    except:
        print ("NotFound")
        mrts_lat.append(0.0)
        mrts_long.append(0.0)
print(mrts_lat )
print(mrts_long )

'''

'\nmrts_lat = []\nmrts_long = []\n# Initialize Nominatim API\ngeolocator=ArcGIS()\n#geolocator = Nominatim(user_agent="MyApp")\n#geolocator = Nominatim(user_agent="geoapiExercises")\n#geolocator = Nominatim(user_agent="Geopy Library")\nfor i in range(0, len(list_of_mrt)):\n    query_address = list_of_mrt[i]\n    query_addressSplit = query_address.split(" ")\n    address = "+".join(query_addressSplit)\n    location = geolocator.geocode(address)\n    try:\n        if location.latitude < 2.0 and location.latitude > 1.0:\n            formattedLat = "{:.6f}".format(location.latitude)\n        else:\n            formattedLat =0.0\n        if location.longitude > 102.0 and location.longitude < 104.0 :\n            formattedLon = "{:.6f}".format(location.longitude)\n        else:\n            formattedLon =0.0\n        \n        print (str(query_address)+",Lat: "+str(formattedLat) +" Long: "+str(formattedLon))\n        mrts_lat.append(float(formattedLat))\n        mrts_long.append(float(format

# As the geocoded data is not so accurate and there are so many missing latitude and Longitude values for many mrts station which does not serve the purpose at all, here the alternate approach is to read a csv from externally to get all the values without missing any

In [31]:
df_mrt_stations = pd.read_csv('csv_data\\External_csv_Files\\mrts_coOrdinates.csv')
df_mrt_stations

MRT  latitude   longitude
0        Jurong East MRT Station  1.333295  103.742154
1        Bukit Batok MRT Station  1.348711  103.749335
2       Bukit Gombak MRT Station  1.358612  103.751791
3      Choa Chu Kang MRT Station  1.385363  103.744371
4            Yew Tee MRT Station  1.397580  103.747459
..                           ...       ...         ...
117    Tampines West MRT Station  1.345634  103.938492
118    Tampines East MRT Station  1.356191  103.954634
119     Upper Changi MRT Station  1.341740  103.961473
120  Woodlands North MRT Station  1.447782  103.785136
121  Woodlands South MRT Station  1.427488  103.792730

[122 rows x 3 columns]

In [32]:
address_list = df_forfeatureEng['address'].unique()      

In [33]:
len(address_list)

9841

In [34]:
# Resale Property Address Lat and Long Value Extraction from geoPy module
# the code block is comented to prevent the recursive scraping of geodata
'''
adress_Lst = []
address_lat = []
address_long = []
# Initialize Nominatim API
#geolocator=ArcGIS()
#geolocator = Nominatim(user_agent="MyApp")
#geolocator = Nominatim(user_agent="geoapiExercises")
geolocator = Nominatim(user_agent="Geopy Library",timeout=15)
for i in range(0, len(address_list)):
    query_address = address_list[i]
    location = geolocator.geocode(query_address)
    try:
        if location.latitude < 2.5 and location.latitude > 0.7:
            formattedLat = "{:.6f}".format(location.latitude)
        else:
            formattedLat =0.0
        if location.longitude > 101.0 and location.longitude < 106.0 :
            formattedLon = "{:.6f}".format(location.longitude)
        else:
            formattedLon =0.0
        print (i ,str(query_address)+",Lat: "+str(formattedLat) +" Long: "+str(formattedLon))
        adress_Lst.append(query_address)
        address_lat.append(float(formattedLat))
        address_long.append(float(formattedLon))

    except:
        print (i ,str("Not Found")+",Lat: "+str(0.0) +" Long: "+str(0.0))
        address_lat.append(0.0)
        address_long.append(0.0)
        adress_Lst.append(query_address)
#print(i )
#print(mrts_long )

df_addressCoordinates = pd.DataFrame({
    'latitude': address_lat,
    'longitude': address_long,
    'address': adress_Lst
})
#len(df_addressCoordinates)

df_addressCoordinates.to_csv('csv_data\\df_addressCoordinates.csv',index=False)

'''

'\nadress_Lst = []\naddress_lat = []\naddress_long = []\n# Initialize Nominatim API\n#geolocator=ArcGIS()\n#geolocator = Nominatim(user_agent="MyApp")\n#geolocator = Nominatim(user_agent="geoapiExercises")\ngeolocator = Nominatim(user_agent="Geopy Library",timeout=15)\nfor i in range(0, len(address_list)):\n    query_address = address_list[i]\n    location = geolocator.geocode(query_address)\n    try:\n        if location.latitude < 2.5 and location.latitude > 0.7:\n            formattedLat = "{:.6f}".format(location.latitude)\n        else:\n            formattedLat =0.0\n        if location.longitude > 101.0 and location.longitude < 106.0 :\n            formattedLon = "{:.6f}".format(location.longitude)\n        else:\n            formattedLon =0.0\n        print (i ,str(query_address)+",Lat: "+str(formattedLat) +" Long: "+str(formattedLon))\n        adress_Lst.append(query_address)\n        address_lat.append(float(formattedLat))\n        address_long.append(float(formattedLon))\n\n

In [35]:
df_addressCoordinatesCsv= pd.read_csv('csv_data\\df_addressCoordinates.csv')

In [36]:
df_addressCoordinatesCsv

latitude   longitude                 address
0     1.353828  103.863739    309 ANG MO KIO AVE 1
1     1.366209  103.841476    216 ANG MO KIO AVE 1
2     1.369205  103.841719    211 ANG MO KIO AVE 3
3     1.368467  103.844154    202 ANG MO KIO AVE 3
4     1.369429  103.861322    235 ANG MO KIO AVE 3
...        ...         ...                     ...
9836  1.445107  103.829278      118B CANBERRA CRES
9837  1.390917  103.873718  455A SENGKANG WEST AVE
9838  1.444610  103.829628      118A CANBERRA CRES
9839  1.444683  103.829890      117B CANBERRA CRES
9840  0.000000    0.000000       516 WEST COAST RD

[9841 rows x 3 columns]

In [37]:
df_addressCoordinatesCsv = df_addressCoordinatesCsv[df_addressCoordinatesCsv.address != "Not Found"]
df_addressCoordinatesCsv = df_addressCoordinatesCsv[df_addressCoordinatesCsv.latitude != 0.000000]
df_addressCoordinatesCsv

latitude   longitude                 address
0     1.353828  103.863739    309 ANG MO KIO AVE 1
1     1.366209  103.841476    216 ANG MO KIO AVE 1
2     1.369205  103.841719    211 ANG MO KIO AVE 3
3     1.368467  103.844154    202 ANG MO KIO AVE 3
4     1.369429  103.861322    235 ANG MO KIO AVE 3
...        ...         ...                     ...
9835  1.323319  103.889449          83A CIRCUIT RD
9836  1.445107  103.829278      118B CANBERRA CRES
9837  1.390917  103.873718  455A SENGKANG WEST AVE
9838  1.444610  103.829628      118A CANBERRA CRES
9839  1.444683  103.829890      117B CANBERRA CRES

[8203 rows x 3 columns]

In [38]:
df_addressCoordinatesCsv.describe()

latitude    longitude
count  8203.000000  8203.000000
mean      1.370490   103.842798
std       0.042016     0.076443
min       1.270369   102.967208
25%       1.343477   103.778193
50%       1.370220   103.848313
75%       1.396435   103.901185
max       2.012057   103.987586

# as the geocoding is not accurate and there are so many missing latitude and Longitude values for Address and we able to retain only 8203 rows from 9841 rows ,here we lost 1638 rows ,even the remaining row data is out of range in min - max  of lattitude and longitude so here the alternate approach is to read a csv from external source to get all the values without missing any for the accurate Predictions .

In [39]:
# gathering the accurate lattitte and longitude valus from a reliable data with maximum unique address count from a external csv file source.
df_addressCoorExtCSV= pd.read_csv('csv_data\\External_csv_Files\\best_addressCoordinates.csv')
df_addressCoorExtCSV

latitude   longitude                address postal_code
0     1.362005  103.853880  406 ANG MO KIO AVE 10      560406
1     1.370966  103.838202   108 ANG MO KIO AVE 4      560108
2     1.380709  103.835368   602 ANG MO KIO AVE 5      560602
3     1.366201  103.857201  465 ANG MO KIO AVE 10      560465
4     1.381041  103.835132   601 ANG MO KIO AVE 5      560601
...        ...         ...                    ...         ...
9526  1.355461  103.937119    860B TAMPINES AVE 5      522860
9527  1.439166  103.779977     121 MARSILING RISE      730121
9528  1.421422  103.839249       659 YISHUN AVE 4      760659
9529  1.377148  103.774024         271 BANGKIT RD      670271
9530  1.307646  103.860269          802 FRENCH RD      200802

[9531 rows x 4 columns]

In [40]:
#to check max and min of latitude and longitude
df_addressCoorExtCSV.describe()

latitude    longitude
count  9531.000000  9531.000000
mean      1.366758   103.840374
std       0.041290     0.074294
min       1.270380   103.685228
25%       1.339186   103.773286
50%       1.364472   103.845869
75%       1.392391   103.899297
max       1.457071   103.987463

In [41]:
df_add_Mrt_cbd_inf = df_addressCoorExtCSV.copy()

In [42]:
#creating list for latitude and longitude for Property address and MRTS 
lst_add_lat = df_add_Mrt_cbd_inf['latitude']
lst_add_long = df_add_Mrt_cbd_inf['longitude']
lst_mrt_lat = df_mrt_stations['latitude']
lst_mrt_long = df_mrt_stations['longitude']

In [43]:
# Combining the Latitude and Longitude  List together 
PropAdd_coOrd_lst = []
MRTS_coOrd_lst = []

for lat, long in zip(lst_add_lat, lst_add_long):
    PropAdd_coOrd_lst.append((lat,long))
for lat, long in zip(lst_mrt_lat, lst_mrt_long):
    MRTS_coOrd_lst.append((lat, long))

# Getting the proximity Values with geopy geodesic

In [44]:
# Distance to nearest MRT
from geopy.distance import geodesic

list_of_dist_mrt = []
min_dist_mrt = []

for origin in PropAdd_coOrd_lst:
    for destination in range(0, len(MRTS_coOrd_lst)):
        list_of_dist_mrt.append(geodesic(origin,MRTS_coOrd_lst[destination]).meters)
    shortest = (min(list_of_dist_mrt))
    min_dist_mrt.append(shortest)
    list_of_dist_mrt.clear()

In [45]:
# Distance from CDB
cbd_dist = []

for origin in PropAdd_coOrd_lst:
    cbd_dist.append(geodesic(origin,(1.2830, 103.8513)).meters) #CBD coordinates

In [46]:
# Put MRT and CBD distance together
df_add_Mrt_cbd_inf['cbd_Dist'] = cbd_dist
df_add_Mrt_cbd_inf['mrt_minDist'] = min_dist_mrt

In [47]:
df_add_Mrt_cbd_inf

latitude   longitude                address postal_code      cbd_Dist  \
0     1.362005  103.853880  406 ANG MO KIO AVE 10      560406   8740.633473   
1     1.370966  103.838202   108 ANG MO KIO AVE 4      560108   9835.486576   
2     1.380709  103.835368   602 ANG MO KIO AVE 5      560602  10948.659905   
3     1.366201  103.857201  465 ANG MO KIO AVE 10      560465   9223.353648   
4     1.381041  103.835132   601 ANG MO KIO AVE 5      560601  10989.223687   
...        ...         ...                    ...         ...           ...   
9526  1.355461  103.937119    860B TAMPINES AVE 5      522860  12466.578115   
9527  1.439166  103.779977     121 MARSILING RISE      730121  19004.920891   
9528  1.421422  103.839249       659 YISHUN AVE 4      760659  15364.619802   
9529  1.377148  103.774024         271 BANGKIT RD      670271  13503.256987   
9530  1.307646  103.860269          802 FRENCH RD      200802   2902.265148   

      mrt_minDist  
0      999.941595  
1     1268.958162  
2     1076.298560  
3      945.371837  
4     1099.505343  
...           ...  
9526  1097.338326  
9527   950.372898  
9528   828.374098  
9529  1339.847514  
9530   279.630252  

[9531 rows x 6 columns]

In [48]:
df_add_Mrt_cbd_inf.to_csv('csv_data\\mrts_cbd_Dist_onAddress_Features.csv',index=False)

In [49]:
df_coordinates = pd.read_csv('csv_data\\mrts_cbd_Dist_onAddress_Features.csv')
df_coordinates

latitude   longitude                address postal_code      cbd_Dist  \
0     1.362005  103.853880  406 ANG MO KIO AVE 10      560406   8740.633473   
1     1.370966  103.838202   108 ANG MO KIO AVE 4      560108   9835.486576   
2     1.380709  103.835368   602 ANG MO KIO AVE 5      560602  10948.659905   
3     1.366201  103.857201  465 ANG MO KIO AVE 10      560465   9223.353648   
4     1.381041  103.835132   601 ANG MO KIO AVE 5      560601  10989.223687   
...        ...         ...                    ...         ...           ...   
9526  1.355461  103.937119    860B TAMPINES AVE 5      522860  12466.578115   
9527  1.439166  103.779977     121 MARSILING RISE      730121  19004.920891   
9528  1.421422  103.839249       659 YISHUN AVE 4      760659  15364.619802   
9529  1.377148  103.774024         271 BANGKIT RD      670271  13503.256987   
9530  1.307646  103.860269          802 FRENCH RD      200802   2902.265148   

      mrt_minDist  
0      999.941595  
1     1268.958162  
2     1076.298560  
3      945.371837  
4     1099.505343  
...           ...  
9526  1097.338326  
9527   950.372898  
9528   828.374098  
9529  1339.847514  
9530   279.630252  

[9531 rows x 6 columns]

In [50]:
colList=df_coordinates.columns
for col in colList:
    print(col+":" +"NaN_Count=" +str( df_coordinates[col].isnull().sum()))
print("")

latitude:NaN_Count=0
longitude:NaN_Count=0
address:NaN_Count=0
postal_code:NaN_Count=0
cbd_Dist:NaN_Count=0
mrt_minDist:NaN_Count=0



In [55]:
df_coordinates.to_csv('csv_data\\mrts_cbd_Dist_onMainDF_ToML.csv',index=False)